### ISSUE TITLE
Add "generate bibtex from identifier" button in toolbar

### ISSUE THREAD

**Issue by tobiasdiez** [JabRef/jabref#4183]
There should be a button in the toolbar which allows to generate a new entry from a given identifier. We could just open the "New entry" dialog on button click, or, what might be more user friendly, open a flyout popup that only contains the lower part of the "New entry" dialog.

**Comment by chrisjmendoza** [comment]
I'm working on this. Already implemented a "New Entry" button next to new article. Working on implementing fly-out.

**Comment by atulim** [comment]
@chrisjmendoza r u still working on this?

**Comment by koppor** [comment]
Refs #550

**Comment by CarodePourtales** [comment]
Hello, 
We have to work on an open source project with a group of 6 students for a course in an engineering school and we chose to work on your project. We also choose this issue. Could you gave us further details  on what you want ? Thank you a lot !

**Comment by tobiasdiez** [comment]
@CarodePourtales Thanks for your interest!

The feature should work as follows:
- Add a button in the toolbar
- When the user clicks this button, then a flyout should be displayed with a textbox where the user could insert an ID (DOI, ISBN, ...).
- JabRef then tries to determine the type of the identifier, fetches the corresponding bibliographic information and then creates a new entry with this data.

Implementation wise, the last step is already implemented in https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/gui/EntryTypeViewModel.java#L117 apart from the automatic determination of the type of the identifier. Let us know if you need more details and/or if there are still things that are unclear. It would be nice if you could open a pull request very early in your development to outline your ideas and show prototypes.

**Comment by CarodePourtales** [comment]
Thanks but isn't the issue resolved like the picture shows or do you want something more ? 
![Capture d’écran 2019-11-27 à 14 30 27](https://user-images.githubusercontent.com/49019500/69727278-823fc900-1122-11ea-9686-29653f0028b3.png)

**Comment by tobiasdiez** [comment]
Yes, the feature in itself kind of exists. The idea was to move the functionality from the lower part of this dialog to the button in the toolbar. In addition, the `ID type` combobox should be replaced by an automatic detection of the identifier type (if possible). 

So it's a bit of work, but not that involved (depending on how much time you are supposed to invest, it might make sense to look at another issue too).

**Comment by koppor** [comment]
Please also read along the discussion at https://github.com/JabRef/jabref/issues/550#issuecomment-397175552. It should provide an idea of the user requirements.

**Comment by CarodePourtales** [comment]
Hello, I was wondering what a crossref id looks like, because I am using Regex to know what IDBasedFetcher I should use to parse the ID the user enters but I don't know what the user should enter to use a Crossref fetcher.
Thanks

**Comment by CarodePourtales** [comment]
It looks like that : for Doi, diva and many other fetchers it works because the pattern is very particular and I can use REGEX, but I don't understand Crossref
![Capture d’écran 2019-12-08 à 16 32 43](https://user-images.githubusercontent.com/49019500/70391709-6bc62700-19d8-11ea-83c6-2a0e21539e1e.png)

**Comment by CarodePourtales** [comment]
I will of course change the look of the string showed but it is just a prototype for the moment

**Comment by matthiasgeiger** [comment]
👍 

Crossref is a fetcher for DOI data (see https://search.crossref.org/)

**Comment by tmrd993** [comment]
I'd like to play around with this but I don't want to be assigned yet. How would we check if an entered ID is valid? Implement regular expressions for every fetcher? Try every fetcher until a result pops up?

**Comment by tobiasdiez** [comment]
Nice to hear! 

The most versatile approach would be to extend the [IdentifierParser](https://github.com/JabRef/jabref/blob/master/src/main/java/org/jabref/logic/importer/util/IdentifierParser.java) by a method that tries to parse the string into one of the Identifier classes. This is a bit similar to the existing https://github.com/JabRef/jabref/blob/a0284e1c61cb465bdd9069e17c61f181028a0bba/src/main/java/org/jabref/logic/importer/util/IdentifierParser.java#L25-L38, except that we try all parsers and not select one based on a field. Given the parsed identifier, one can then use the following method to get the corresponding fetcher:
https://github.com/JabRef/jabref/blob/a0284e1c61cb465bdd9069e17c61f181028a0bba/src/main/java/org/jabref/logic/importer/WebFetchers.java#L67-L74

(As of now this only covers DOIs, but that should be relatively straightforward to extend).

**Comment by tmrd993** [comment]
I can see that some of the identifier classes have methods that check the validity of the given String id. Shouldn't that be a part of the Identifier interface? Parsing the string into one of the identifier classes would be easier that way. Also, not every fetcher has a corresponding Identifier class as far as I can see. Am I correct in assuming those should be implemented or do they all fall within the available Identifier classes. For example, Crossref and DOI-fetcher both match with the DOI-Identifier but there is no Identifier class for medline or diva.

**Comment by tobiasdiez** [comment]
> have methods that check the validity of the given String id. Shouldn't that be a part of the Identifier interface?

These are static methods, and thus cannot be lifted to the interface. Maybe it would make sense to introduce a `IdentifierParser` interface?!

> Also, not every fetcher has a corresponding Identifier class as far as I can see.

Right, that's still work in progress. I would say you can ignore this problem for now and add the button in the toolbar for all fetcher that do have an identifier

**Comment by tmrd993** [comment]
> These are static methods, and thus cannot be lifted to the interface. Maybe it would make sense to introduce a `IdentifierParser` interface in place of `Function<String, Optional<? extends Identifier>>`?
> 

Yes, I think that makes sense but there is already a class called IdentifierParser and I'm drawing a blank on what to call that one instead if we are creating an interface with that name. Looking at the class, it seems more like a field parser to me?

**Comment by tobiasdiez** [comment]
You can rename this class to `IdentifierParsers` (with s at the end) since it's a factory for the IdentifierParser. We have a similar convention already for `WebFetcher(s)`

**Comment by tmrd993** [comment]
Another question, can I use a dialogpane instead of a flyout for now? I'm not seeing a jabref component (or a javafx component) that I can use for that, should that be implemented from scratch? Anyways, I implemented it the same way the new entry view works for a start, by opening a new dialogpane that returns an entrytype. This is what it looks like right now  (It also only works for DOI (from CrossRef), Arxiv and ISBN fetchers at the moment as those are the only types that have implemented Identifier classes):

![newIdEntry](https://user-images.githubusercontent.com/47194464/110522988-d4d3a780-8111-11eb-8ac3-cfb5733e62dd.gif)

**Comment by tobiasdiez** [comment]
That looks already pretty nice,  thanks for your work!

For the flyout, you can use the popover control. As an example:
https://github.com/JabRef/jabref/blob/c7f2671730f28fb75af823f574ba7d27106c1b06/src/main/java/org/jabref/gui/JabRefFrame.java#L941-L945

**Comment by koppor** [comment]
@tmrd993 May I ask for an update? Is there some chance to get this into JabRef?

**Comment by koppor** [comment]
For the implementation, please also investigate #550. That request asks that JabRef also recognizes `10.1016/j.jrmge.2015.08.004` as DOI and `978-3-86680-192-9` as ISBN.

This issue describes that the lower part of the "new entry" dialog should be duplicated to a toolbar **submenu** only for an id:

![grafik](https://user-images.githubusercontent.com/1366654/123559816-ae0fae80-d79e-11eb-928e-f0138be6f6bd.png)

Thus, this is not just a request to add some new dialog, but to work with JavaFX's toolbar magic to create an input possibility directly at a toolbar button.

**Comment by tmrd993** [comment]
I had stopped working on it because of finals, I'll get back on it if I find the time but I'm not claiming it or anything, if someone else wants to solve this feel free to.

**Comment by tmrd993** [comment]
> 
> This issue describes that the lower part of the "new entry" dialog should be duplicated to a toolbar **submenu** only for an id:
>
> Thus, this is not just a request to add some new dialog, but to work with JavaFX's toolbar magic to create an input possibility directly at a toolbar button.

I'm not sure I understand the requirements, is it supposed to be a Menu object inside the toolbar that opens up a text input for the id? Is there some graphical example of this?

This is how I pictured it:

![3zhPH](https://user-images.githubusercontent.com/47194464/124616777-a7f09080-de76-11eb-9772-fa8cd2bf5ede.png)

**Comment by koppor** [comment]
Looks good. The implementation should try different Ids and try to fetch it. --> The user should not be required to input the Id type. Only the Id (as you shown it).

The position should be below the button (as I interpret your screenshot)

![grafik](https://user-images.githubusercontent.com/1366654/125522339-4da067da-d566-4cad-b215-8ead7d212605.png)
